MAINT: Complete Conversion to Pytest Idiom

gfyoung opened the pull request:
* Convert all test setup/teardown to `pytest` idiom
* Remove any classes that inherit from `unittest.TestCase`
* Update documentation that we are still using `tm.TestCase`

Closes #15990.

`pytest` idiom for test setup/teardown can be found <a href="https://docs.pytest.org/en/2.7.3/xunit_setup.html">here</a>.

gfyoung committed:
MAINT: Convert test setup/teardown to pytest idiom

* tm.TestCase now just inherits from object
* setUpClass renamed to setup_class
* tearDownClass renamed to teardown_class
* setUp renamed to setup_method
* tearDown renamed to teardown_method

gfyoung committed:
MAINT: Remove unittest.TestCase from testing

gfyoung committed:
DOC: Update documentation for TestCase usage

tm.TestCase no longer follows the nosetest idiom,
so it is here to stay, so update the documentation
to say that we are using it still.

Closes gh-15990.

gfyoung commented:
@jreback : This test failure <a href="https://travis-ci.org/pandas-dev/pandas/jobs/228021399">here</a> makes no sense to me at all.  I can't reproduce this when I walk through the test (though it replicates when I run the tests locally).

jreback commented:
```
/home/travis/build/pandas-dev/pandas/pandas/tests/frame/test_query_eval.py yield tests are deprecated, and scheduled to be removed in pytest 4.0
```

gfyoung commented:
@jreback : I don't really follow.  What's the point of this comment?

jreback commented:
that shouldn't happen, that means a fixture is being used incorrectly.

jreback commented:
hmm, that could be. so you can relax / eliminate that test then. prob has to do with ordering of the cache insertions.

gfyoung commented:
But that cache insertion should never have happened though.  I'm surprised that this only surfaced now (don't really see how my change would have caused it).

gfyoung commented:
Something is going wrong with `round_trip_pickle`.  You can replicate here:
~~~python
>>> dtype = IntervalDtype("int64")
>>> IntervalDtype._cache
{'int64': interval[int64]}
>>>
>>> result = tm.round_trip_pickle(dtype)
>>> IntervalDtype._cache
{'int64': interval[int64], 'None': interval[int64]}
~~~

gfyoung commented:
Something goes wrong when you call `round_trip_pickle`.  You can replicate here:
~~~python
>>> from pandas.core.dtypes.dtypes import IntervalDtype
>>> import pandas.util.testing as tm
>>>
>>> dtype = IntervalDtype("int64")
>>> IntervalDtype._cache
{'int64': interval[int64]}
>>>
>>> none_dtype = IntervalDtype("interval")
>>> IntervalDtype._cache
{'int64': interval[int64], 'None': interval}
>>>
>>> result = tm.round_trip_pickle(dtype)
>>> IntervalDtype._cache  # what!?
{'int64': interval[int64], 'None': interval[int64]}
~~~
Something goes wrong when you call `round_trip_pickle`.  You can replicate here:
~~~python
>>> from pandas.core.dtypes.dtypes import IntervalDtype
>>> import pandas.util.testing as tm
>>>
>>> dtype = IntervalDtype("int64")
>>> IntervalDtype._cache
{'int64': interval[int64]}
>>>
>>> none_dtype = IntervalDtype("interval")
>>> IntervalDtype._cache
{'int64': interval[int64], 'None': interval}
>>>
>>> result = tm.round_trip_pickle(dtype)
>>> IntervalDtype._cache  # what!?
{'int64': interval[int64], 'None': interval[int64]}
~~~
A closer look indicates that there is aliasing going on in the cache.  That's why the value mutates so mysteriously, which happen the pickled object is reloaded.

gfyoung commented:
@jreback : What was the rationale for using a `_cache` ?  Is it possible that we can just remove this caching given that this interferes with pickling?

gfyoung commented:
The other option is that we add a `copy` parameter to the constructor, so that we can return copies and avoid this kind of aliasing.

gfyoung commented:
@jreback : What was the rationale for having a `_cache` attribute (this type of construction seems particular to just these duck-typed `dtype`)?  Is it possible that we can just remove this caching given that this interferes with pickling?

The other option is that we add a `copy` parameter to the constructor, so that we can return copies and avoid this kind of aliasing.

jreback commented:
i will fix 
the cache is important here
it shouldn't be serializing though

gfyoung commented:
Sorry, I should have said "unpickling," but okay, I will await a patch then for the aliasing then.

jreback commented:
FYI merging #16207 shortly. you can pick up ``pandas/tests/dtypes/test_dtypes.py`` from master as I fixed this for pytest.

gfyoung commented:
@jreback : Circle is failing, and I can't reproduce it.  It seems like there is some kind of weird `matplotlib` import failure that shouldn't be happening, but I have no idea why it is happening (can't reproduce locally with and without `matplotlib` installed).

jreback commented:
not sure. you should try rolling back parts of the change until this work.

gfyoung committed:
TST: Patch Circle matplotlib failure

The tm.mplskip decorator was breaking on Circle,
so this commit removes the decorator and replaces it
with direct function calls to check for matplotlib.

gfyoung commented:
@jreback : I managed to patch the errors in Circle, so everything is green now.  PTAL.

gfyoung commented:
@jreback : I managed to patch the errors in Circle, so everything is green now.  PTAL.

TomAugspurger commented on pandas/util/testing.py:
```diff
@@ -299,18 +293,25 @@ def mplskip(cls):
     return cls
 
 
+def _skip_module_if_no_mpl():
```
This was for the matplotlib failures, correct? `tm.mplskip` is used on one class in `pandas/tests/io/formats/test_style.py`. From what I can see that's the only place it's used.

If you want, you could add a `pytest.importorskip('matplotlib')` on that test and delete `mplskip` altogether.

gfyoung replied:
That is correct.  I don't see why not.  Let me try and see what happens.

TomAugspurger commented:
@gfyoung I'll have a commit in a minute that fixes those yield warnings. Basically converting

```
-    def test_query_with_named_multiindex(self):
-        for parser, engine in product(['pandas'], ENGINES):
-            yield self.check_query_with_named_multiindex, parser, engine
-
-    def check_query_with_named_multiindex(self, parser, engine):
+    @pytest.mark.parametrize('parser', ['pandas'])
+    @pytest.mark.parametrize('engine', ENGINES)
+    def test_query_with_unnamed_multiindex(self, parser, engine):
 ```

In a bunch of places. Mind if I push that here?

gfyoung commented:
@TomAugspurger : Certainly.  Perhaps wait until the tests pass though?

TomAugspurger commented:
@gfyoung they all passed. Pushing that extra commit now.

gfyoung commented:
@TomAugspurger : Awesome.  Nice that this PR has been able to pick several other things beyond its original purpose (cache bug, `pytest` warnings...) :smile:

TomAugspurger committed:
TST: Replace yield-based tests in test_query_eval

gfyoung commented on pandas/tests/frame/test_query_eval.py:
```diff
@@ -165,7 +164,9 @@ def test_eval_resolvers_as_list(self):
 class TestDataFrameQueryWithMultiIndex(tm.TestCase):
 
-    def check_query_with_named_multiindex(self, parser, engine):
+    @pytest.mark.parametrize('parser', ['pandas'])
```
I wonder if a comment here explaining why we're parametrizing instead of just hardcoding `parser` would be useful.

TomAugspurger replied:
Yeah I thought that too. I went with parametrizing simply because it's symmetric with the other ones where we parametrize over all the `PARSERS`. I don't feel strongly about that, if you prefer the other way.

gfyoung replied:
True.  I suspected that was the rationale.  I don't mind too much at all.  It helps make the test more flexible that way.  However, it may not be so for other developers, that's all.

TomAugspurger replied:
Misread you earlier. Added a comment 👍

jreback commented on doc/source/contributing.rst:
```diff
@@ -633,12 +633,7 @@ framework that will facilitate testing and developing. Thus, instead of writing
-
-    class TestReallyCoolFeature(object):
-        ....
+using test classes allows for better organization. If test classes are used, make sure that they do inherit from ``tm.TestCase``.
```
just remove this entirely

gfyoung replied:
Done.

jreback commented on pandas/tests/frame/test_query_eval.py:
```diff
@@ -165,7 +164,10 @@ def test_eval_resolvers_as_list(self):
 class TestDataFrameQueryWithMultiIndex(tm.TestCase):
 
-    def check_query_with_named_multiindex(self, parser, engine):
+    # parametrize single parser, for symmetry with other tests
```
this is better done with fixtures no? these will then automatically get the cartesian product. now you *may* want to parameterize when you *don't* want that of course (e.g. in this example here you can parameterize on parser, but leave engines as the fixture).

e.g.
```
@pytes.fixture(params=ENGINES)
def engine(request):
    return request.param
```
and same for parser

gfyoung replied:
Yep, that makes sense.  Done.

jreback reviewed (changes requested):
in another PR can then remove ``tm.TestCase`` from the base class of test class where it is clearly not needed (which is almost all places, leave it maybe in ``pandas/tests/frame`` and ``pandas/tests/indexing``. prob can remove elsewhere.

gfyoung commented:
> in another PR can then remove tm.TestCase from the base class of test class where it is clearly not needed

The only thing that `tm.TestCase` contributes is raising on chained assignment (part of `setup_class`).  I presume that is to just discourage tests from being written like this.  Could we just remove that (and just make sure to be vigilant about it when writing tests)?

jreback commented:
> The only thing that tm.TestCase contributes is raising on chained assignment (part of setup_class). I presume that is to just discourage tests from being written like this. Could we just remove that (and just make sure to be vigilant about it when writing tests)?

NO, this is the entire point of this. as I said you can remove from everywhere except pandas/tests/frame & pandas/tests/indexing

jreback commented:
> and just make sure to be vigilant about it when writing tests

people are NEVER vigilant, assume that they aim to break every single rule :>

gfyoung commented:
But the default for chained assignment is "warning" - would that not surface in the build logs if tests were incorrectly written with chained assignment?

jreback commented on pandas/tests/frame/test_query_eval.py:
```diff
@@ -308,12 +315,9 @@ def check_query_with_unnamed_multiindex(self, parser, engine):
-    def check_query_with_partially_named_multiindex(self, parser, engine):
+    def test_query_with_partially_named_multiindex(self, parser, engine):
         tm.skip_if_no_ne(engine)
+        skip_if_no_pandas_parser(parser)
```
you can actually do this with a fixture skip, but no big deal really

jreback commented:
ping when green. I have a PR after this one (moving stuff around).

jreback commented:
> But the default for chained assignment is "warning" - would that not surface in the build logs if tests were incorrectly written with chained assignment?

no its not for testing, its raise.

gfyoung commented:
> no its not for testing, its raise.

But that's because you set it that way when you set up each test class.  I meant in general, it's warning by default.

gfyoung commented:
@jreback : All green and ready to go.

jreback commented:
> But that's because you set it that way when you set up each test class. I meant in general, it's warning by default, hence my question.

but that's exactly the point. if someone writes a *test* that uses chained indexing I WANT it to fail the tests. Sure we can look for it, but its best to automatically catch it.

jreback commented:
thanks for this.

gfyoung commented:
> but that's exactly the point. if someone writes a test that uses chained indexing I WANT it to fail the tests. 

What if we added it to `conftest.py`?

jreback commented:
hmm that might work actually

gfyoung commented:
I just gave it a shot, and I think it might work (here is code snippet)

[test_example.txt](https://github.com/pandas-dev/pandas/files/975149/test_example.txt)

jreback commented:
ok sounds good
